feat: add the w3c test script and server

srjames90 opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/master/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/master/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?
fixes #1363
-

## Short description of the changes
This adds a github action to run the w3c integration tests every time a change is pushed. This works by checking out the w3c test harness found here https://github.com/w3c/trace-context/tree/master/test, and spins up a simple js server that receives request from a client spun up by the test harness and sends requests to a python server spun up by the test harness that then checks if the headers have been propagated in compliance with the w3c spec.

srjames90 commented on .github/workflows/w3c-integration-test.yml:
```diff
@@ -0,0 +1,21 @@
+      - name: Install and Build 🔧
+        run: |
+          npm install --ignore-scripts
+          #npx lerna bootstrap --scope={propagation-tests,@opentelemetry-api} --include-dependencies
+          #npx lerna link ----force-local
```
Anyone know how to use lerna really well? This works locally, but when I try it here in the ubuntu machine, I get ```PluginLoader#load: could not load plugin @opentelemetry/plugin-http of module http. Error: Cannot find module '@opentelemetry/plugin-http'``` in the next step.

srjames90 replied:
I'm currently just running `npm i` in the bash script to get around this.

dyladan replied:
It's because your bootstrap on the line above excludes the http plugin from the bootstrap.

srjames90 commented on scripts/tracecontext-integration-test.sh:
```diff
@@ -0,0 +1,26 @@
+cd ./target/trace-context && git checkout $TRACECONTEXT_GIT_TAG && cd -
+pip3 install setuptools;
+pip3 install aiohttp; 
+(cd validation_server; npm i)
```
This is the workaround for https://github.com/open-telemetry/opentelemetry-js/pull/1380/files#r464520833, but it's not a real workaround because it's not testing against the latest commit but the latest released version I think

srjames90 commented on validation_server/validation_server.js:
```diff
@@ -0,0 +1,54 @@
+
+// Get the propagator type
+const propagatorTypeArg = myArgs[0];
+const myPropagator =
+  propagatorTypeArg === "b3" ? B3Propagator : HttpTraceContext;
```
Adding this for the future because we can add something similar for b3

dyladan commented on validation_server/package.json:
```diff
@@ -0,0 +1,19 @@
+  "author": "",
+  "license": "ISC",
+  "dependencies": {
+    "@opentelemetry/api": "^0.10.1",
```
We are now on 10.2

dyladan commented on validation_server/package.json:
```diff
@@ -0,0 +1,19 @@
+  "dependencies": {
+    "@opentelemetry/api": "^0.10.1",
+    "@opentelemetry/node": "^0.10.1",
+    "@opentelemetry/plugin-express": "^0.8.0",
```
You are depending on a lot of autoinstrumentation modules for what should be a very basic test. I would skip everything except for `@opentelemetry/{api,core,tracing}`, and call the propagation and tracing apis manually.

dyladan commented on validation_server/validation_server.js:
```diff
@@ -0,0 +1,54 @@
+const axios = require("axios");
+const { B3Propagator } = require("@opentelemetry/core");
```
you require core twice. B3 is outside of the scope currently, I would remove it for now.

dyladan commented on validation_server/validation_server.js:
```diff
@@ -0,0 +1,54 @@
+const axios = require("axios");
+const { B3Propagator } = require("@opentelemetry/core");
+const { HttpTraceContext } = require("@opentelemetry/core");
+const { NodeTracerProvider } = require("@opentelemetry/node");
```
Nothing here is node specific. I would use the basic tracer provider provided in the `tracing` package and call propagation APIs manually.

srjames90 commented:
https://github.com/open-telemetry/opentelemetry-js/pull/1380#discussion_r465249667
@dyladan Yeah, I had tried including the http-plugin with bootstrap, but it still said the same. I'll try again once I make the changes you suggested. Thanks.

dyladan commented:
> [#1380 (comment)](https://github.com/open-telemetry/opentelemetry-js/pull/1380#discussion_r465249667)
> @dyladan Yeah, I had tried including the http-plugin with bootstrap, but it still said the same. I'll try again once I make the changes you suggested. Thanks.

I would suggest removing the http plugin entirely, and just using the propagation API. This will drastically simplify the test.

srjames90 commented:
> > [#1380 (comment)](https://github.com/open-telemetry/opentelemetry-js/pull/1380#discussion_r465249667)
> > @dyladan Yeah, I had tried including the http-plugin with bootstrap, but it still said the same. I'll try again once I make the changes you suggested. Thanks.
> 
> I would suggest removing the http plugin entirely, and just using the propagation API. This will drastically simplify the test.

Cool, sounds good.

dyladan commented on validation_server/validation_server.js:
```diff
@@ -1,27 +1,33 @@
-const myPropagator =
-  propagatorTypeArg === "b3" ? B3Propagator : HttpTraceContext;
+  setExtractedSpanContext,
+} = require("@opentelemetry/core");
```
You require things from core, tracing, and context-base, but none of them are in your dependencies.

dyladan commented on validation_server/validation_server.js:
```diff
@@ -0,0 +1,82 @@
+
+// Mount our demo route
+app.post("/verify-tracecontext", (req, res) => {
+  const context = myPropagator.extract(
```
Use the api here, not the propagator directly.

`propagation.extract(req.headers)`

dyladan commented:
Based on the action output, it looks like we might actually just be failing the strict test harness. I'll take a look in the morning.

srjames90 commented:
> Based on the action output, it looks like we might actually just be failing the strict test harness. I'll take a look in the morning.

The bash script sets the strictness level to 1, so the harness skips the stricter tests.

dyladan commented:
You have a couple issues that I noticed.

1. your validation server package.json requires 0.10.2, but your branch is on version 0.10.1. This means you're getting the npm version, not the local version when you bootstrap.
2. You are severely out of date. The traceparent handling has changed recently. please merge master into your branch
3. The folder is called `validation-tests` but the package name is `propagation-tests`
4. The version of your tests package does not match the version in lerna.json and other packages. All versions should stay in sync for the best working of lerna
5. on your branch we appear to fail some tests, but I believe many of them are fixed in the current version
6. your bootstrap scope includes api specifically, but also includes all dependencies. api is a dependency of your tests so you can just do `npx lerna bootstrap --scope=propagation-tests --include-dependencies`
7. the validation-server package.json should have `private: true` to prevent us from accidentally publishing it as an npm package

dyladan commented:
Can you try bumping the strictness down to level 2?

srjames90 commented:
> Can you try bumping the strictness down to level 2?

level 2 is even more strict. It adds more tests cases.

dyladan commented:
No, I thought there was a less strict mode. In the meeting today, @michaelgoin volunteered to look into the test failures. As far as this PR is concerned, I think the tests failing is not a problem. The PR is only to run the test suite.

dyladan commented on .github/workflows/w3c-integration-test.yml:
```diff
@@ -0,0 +1,20 @@
+
+      - name: Run W3C Test harness
+        run: |
+          chmod 755 ./scripts/tracecontext-integration-test.sh
```
Should not have to chmod in the test. Executable bit is tracked by git. If you make your local executable, you should be able to check that in as a change.

srjames90 replied:
cool, thanks, will change.

srjames90 commented:
> No, I thought there was a less strict mode. In the meeting today, @michaelgoin volunteered to look into the test failures. As far as this PR is concerned, I think the tests failing is not a problem. The PR is only to run the test suite.

Looking at the source of the w3c harness, it excludes tests like this ```@unittest.skipIf(STRICT_LEVEL < 2, "strict")```, but it does allow you to run specific test cases and suites instead.

srjames90 commented:
> You have a couple issues that I noticed.
> 
> 1. your validation server package.json requires 0.10.2, but your branch is on version 0.10.1. This means you're getting the npm version, not the local version when you bootstrap.
> 2. You are severely out of date. The traceparent handling has changed recently. please merge master into your branch
> 3. The folder is called `validation-tests` but the package name is `propagation-tests`
> 4. The version of your tests package does not match the version in lerna.json and other packages. All versions should stay in sync for the best working of lerna
> 5. on your branch we appear to fail some tests, but I believe many of them are fixed in the current version
> 6. your bootstrap scope includes api specifically, but also includes all dependencies. api is a dependency of your tests so you can just do `npx lerna bootstrap --scope=propagation-tests --include-dependencies`
> 7. the validation-server package.json should have `private: true` to prevent us from accidentally publishing it as an npm package

Cool, I think I addressed everything. I also renamed the folder and package name to "propagation-validation-server".

obecny commented on propagation-validation-server/package.json:
```diff
@@ -0,0 +1,18 @@
+{
```
I think this package.json should be similar to what we have in packages including the license, author etc. etc

dyladan commented:
Please stop force-pushing changes. A regular commit history with no rebases or history modifications makes it _much_ easier to follow the PR.

srjames90 commented:
> Please stop force-pushing changes. A regular commit history with no rebases or history modifications makes it _much_ easier to follow the PR.

Sorry, I was trying to only force push changes that weren't relevant to the comments in the review, like if I wanted to rerun the action or I forgot a dependency.

dyladan commented:
![image](https://user-images.githubusercontent.com/1612643/89465959-ffefb600-d740-11ea-8e98-7dfbe6bee806.png)

Actions an be re-run with this button :)

srjames90 commented:
> ![image](https://user-images.githubusercontent.com/1612643/89465959-ffefb600-d740-11ea-8e98-7dfbe6bee806.png)
> 
> Actions an be re-run with this button :)

I was blind and now I see lol...wow

srjames90 commented:
> If you register a context manager (I recommended the AsyncHooksContextManager or AsyncLocalStorageContextManager), you should not have to pass the context object manually. You can do something like this:
> 
> ```
> context.with(propagation.extract(req.headers), () => {
>   const span = tracer.startSpan("HTTP POST");
>   tracer.withSpan(span, () => {
>     const headers = {};
>     propagation.inject(headers);
>     // ... and so on
>   });
> });
> ```

Cool, haven't used this before, will try, but I'm wondering if it will make the `.map` code weird since I have to wait for all of the promises to resolves before sending a response.

dyladan commented:
> > If you register a context manager (I recommended the AsyncHooksContextManager or AsyncLocalStorageContextManager), you should not have to pass the context object manually. You can do something like this:
> > ```
> > context.with(propagation.extract(req.headers), () => {
> >   const span = tracer.startSpan("HTTP POST");
> >   tracer.withSpan(span, () => {
> >     const headers = {};
> >     propagation.inject(headers);
> >     // ... and so on
> >   });
> > });
> > ```
> 
> Cool, haven't used this before, will try, but I'm wondering if it will make the `.map` code weird since I have to wait for all of the promises to resolve before sending a response. Will mess around with it a bit today.

If it doesn't work, no problem. We can always merge this and iterate it.

srjames90 commented:
> > > If you register a context manager (I recommended the AsyncHooksContextManager or AsyncLocalStorageContextManager), you should not have to pass the context object manually. You can do something like this:
> > > ```
> > > context.with(propagation.extract(req.headers), () => {
> > >   const span = tracer.startSpan("HTTP POST");
> > >   tracer.withSpan(span, () => {
> > >     const headers = {};
> > >     propagation.inject(headers);
> > >     // ... and so on
> > >   });
> > > });
> > > ```
> > 
> > 
> > Cool, haven't used this before, will try, but I'm wondering if it will make the `.map` code weird since I have to wait for all of the promises to resolve before sending a response. Will mess around with it a bit today.
> 
> If it doesn't work, no problem. We can always merge this and iterate it.

It works, just stored each promise in a variable, and returned that.

michaelgoin commented:
Not to derail this PR but to follow-up: looks like we do have at least a couple spec violations and at least one *should* we don't do (which is less of a big deal). I've only debugged through one, thus far, so could also be a hidden bug impacting the others which may be mostly implemented correct.

I'll prob throw together a PR resolving all the things (STRICT_LEVEL=1) and then we can decide if we have concerns about any of the impacts.

dyladan commented:
> Not to derail this PR but to follow-up: looks like we do have at least a couple spec violations and at least one _should_ we don't do (which is less of a big deal). I've only debugged through one, thus far, so could also be a hidden bug impacting the others which may be mostly implemented correct.
> 
> I'll prob throw together a PR resolving all the things (STRICT_LEVEL=1) and then we can decide if we have concerns about any of the impacts.

I think the failing tests should be solved in a separate PR.

michaelgoin commented:
Agreed. Instead of "prob" I should have said "will" throw together a PR.

michaelgoin commented on integration-tests/tracecontext-integration-test.sh:
```diff
@@ -0,0 +1,26 @@
+# give the app server a little time to start up. Not adding some sort
+# of delay would cause many of the tracecontext tests to fail being
+# unable to connect.
+sleep 1
```
if we find this gets flaky in the future, a health-check approach can be considered. works fine for me, currently.

srjames90 replied:
Fasho, sounds good

mayurkale22 commented on integration-tests/propagation-validation-server/validation-server.js:
```diff
@@ -0,0 +1,68 @@
+context.setGlobalContextManager(new AsyncHooksContextManager());
+
+// Create a provider for activating and tracking spans
+const tracerProvider = new BasicTracerProvider({});
```
```suggestion
const tracerProvider = new BasicTracerProvider();
```

mayurkale22 commented on .github/workflows/w3c-integration-test.yml:
```diff
@@ -0,0 +1,18 @@
+name: Run w3c tests on push
```
Just curious to know, let's say in the future if we decide to extend this workflow to other propagators (such as `B3Propagator`), how would we go about it? Do we need to add new workflows or update this one?

srjames90 replied:
We could do either. Adding the support for that in the server should be easy if we have it accept a command-line argument, which would determine which propagator to use. Then, we can either add a new step in the workflow that tests b3, or add a new workflow.

srjames90 replied:
I was thinking it'd just be a new step.
